compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELogicalObjFifoSplittingUtils.cpp

```diff
@@ -353,39 +342,80 @@ LogicalResult splitLogicalObjectFifos(
   toBeErased.insert(sourceAllocOp);
   toBeErased.insert(sourceObjectFifo);
 
-  SmallVector<OpFoldResult, 4> staticL2AsTargetOffsets =
+  SmallVector<OpFoldResult> staticL2AsTargetOffsets =
       l3ToL2DmaOp.getTargetMixedOffsets();
-  SmallVector<OpFoldResult, 4> staticL2AsTargetSizes =
+  SmallVector<OpFoldResult> staticL2AsTargetSizes =
       l3ToL2DmaOp.getTargetMixedSizes();
-  SmallVector<OpFoldResult, 4> staticL3AsSourceOffsets =
+  SmallVector<OpFoldResult> staticL3AsSourceOffsets =
       l3ToL2DmaOp.getSourceMixedOffsets();
-  SmallVector<OpFoldResult, 4> staticL3AsSourceSizes =
+  SmallVector<OpFoldResult> staticL3AsSourceSizes =
       l3ToL2DmaOp.getSourceMixedSizes();
+
+  LogicalObjectFifoFromMemrefOp l2TargetObjectFifo =
+      l3ToL2DmaOp.getTargetObjectFifo();
+  ArrayRef<int64_t> l2TargetShape =
+      l2TargetObjectFifo.getMemrefType().getShape();
+  if (l2TargetShape.size() != staticL2AsTargetSizes.size()) {
+    LLVM_DEBUG(llvm::dbgs() << "L2 target size should be the same");
+    return failure();
+  }
+
+  // Check if the L3->L2 dma is transposed on the target side.
+  bool dmaTransposeOnSource = true;
+  for (auto [s1, s2] : llvm::zip_equal(l2TargetShape, staticL2AsTargetSizes)) {
+    if (s1 != getConstantIntValue(s2)) {
+      dmaTransposeOnSource = false;
+    }
+  }
+
   OpFoldResult zeroVal = getAsIndexOpFoldResult(context, 0);
   OpFoldResult oneVal = getAsIndexOpFoldResult(context, 1);
-  // Update split dimensions' offset/size for L2 as target and L3 as source. We
-  // can afford to do this here because it's going to be the same for all L3->L2
-  // splits. Here we are setting offset = 0 and size = 1.
-  for (size_t dim : splitDimsForL2) {
-    staticL2AsTargetOffsets[dim] = zeroVal;
-    staticL2AsTargetSizes[dim] = oneVal;
-    staticL3AsSourceOffsets[dim] = zeroVal;
-    staticL3AsSourceSizes[dim] = oneVal;
+
+  if (dmaTransposeOnSource) {
+    // Update split dimensions' offset/size for L2 as target and L3 as source.
+    // We can afford to do this here because it's going to be the same for all
+    // L3->L2 splits. Here we are setting offset = 0 and size = 1.
+    for (size_t dim : splitDimsForL2) {
+      staticL2AsTargetOffsets[dim] = zeroVal;
+      staticL2AsTargetSizes[dim] = oneVal;
+      staticL3AsSourceOffsets[dim] = zeroVal;
+      staticL3AsSourceSizes[dim] = oneVal;
+    }
+  } else {
+    // The L2 target side has transposed dimensions, while the L3 source side
+    // data are continuous and doesn't have "nonSplitDim".
+    // Hardcoded the transposed dimensions for now.
+    const SmallVector<size_t> transposeDim = {0, 2, 1, 3};
```

Comment:
Since we have added a check earlier : `if (l2TargetShape.size() != staticL2AsTargetSizes.size()) `, I'd rather we infer even the dimensions where transpose has taken place inside the loop `for (auto [s1, s2] : llvm::zip_equal(l2TargetShape, staticL2AsTargetSizes))`  itself and get rid of this hardcoding.

Reply:
Well, now I think there could be issues if we blindly compare values. There could be dimensions that have same values, e.g, [2, 32, 32, 32] -> [2, 32, 32, 32] after transposing dim 1 and 2. Then there's no way to tell if the dma is transposed or not.

Reply:
Oh okay. That makes sense. Thanks for the explanation.